### ISSUE TITLE
[mozilla-flame] Append `dtb` file to `boot.img` (qcdt)

### ISSUE THREAD

**Issue by ollieparanoid** [postmarketOS/pmbootstrap#323]
From what @yuvadm told us in the channel, the `flame` comes with its own version of `fastboot`/`mkbootimg`, which appends the `dtb` file at the end of the `boot.img` (and not at the end of the kernel, as it is with all other Android devices). The kernel is part of the `boot.img` file though.

I'm not sure what the best way would be to handle this right now (suggestions welcome!) - @yuvadm, could you point us to the source code of these mozilla specific forks?

**Comment by yuvadm** [comment]
@ollieparanoid Thanks for opening an issue on this topic. Also CC'ing @z3ntu here as well since he provided some initial patches.

Essentially we need to pick a `mkbootimg` that supports the `--dt` flag which is able to append the device tree files to the `boot.img`. This is needed not only for the Flame but also for another device I believe @z3ntu is working on porting.

There are several different patched `mkbootimg` floating around, I'll paste a few that I ran into here, and then we can decide on the merits of each one:

 - https://github.com/efidroid/build/blob/master/tools/mkbootimg
 - https://github.com/LineageOS/android_system_core/blob/cm-14.1/mkbootimg/mkbootimg

Once we figure this out, we need to **conditionally** add the `--dt` flag to `mkbootimg`. An initial breaking patch looks like this:

```diff
--- a/aports/postmarketos-mkinitfs/mkinitfs.sh
+++ b/aports/postmarketos-mkinitfs/mkinitfs.sh
@@ -171,6 +171,7 @@ create_bootimg()
                --ramdisk_offset "${deviceinfo_flash_offset_ramdisk}" \
                --tags_offset "${deviceinfo_flash_offset_tags}" \
                --pagesize "${deviceinfo_flash_pagesize}" \
+               --dt "${kernelfile/initramfs-/dt.img-}" \
                -o "${outfile/initramfs-/boot.img-}"
 }

```

**Comment by z3ntu** [comment]
I have a version of the last patch on my laptop which is non-breaking but I'm on my phone atm

**Comment by z3ntu** [comment]
Also the method used is called qcdt and I would use the lineage source of the tools if available. @m1cha pushed the important changes from efidroid to Lineage

**Comment by yuvadm** [comment]
@z3ntu good to know! Would love to get the updated version whenever you have a chance to post it. Also curious if this worked for me. I'm still running into problems with the flame `dt.img`.

**Comment by ollieparanoid** [comment]
Alright, then I suggest to integrate it the following way:
 - [ ] Replace the `mkbootimg` we have packaged with the version from LineageOS
 - [ ] Modify `linux-mozilla-flame` to save the `dtb` file in `/usr/share/dtb`
 - [ ] Specify the `dtb` file name in the `deviceinfo`
 - [ ] Add a new option: `deviceinfo_bootimg_qcdt="true"`
 - [ ] Modify `mkinitfs.sh` to use the `--dt` flag, when that option is enabled
 - [ ] Test if everything works
 - [ ] Make a PR
 - [ ] Document the new `deviceinfo` option in the deviceinfo reference

**Comment by z3ntu** [comment]
@yuvadm I only got pmOS to boot with a kernel not built via the alpine build process (but normal Android mka zImage)

**Comment by z3ntu** [comment]
A crude version of that patch is https://github.com/z3ntu/pmbootstrap/commit/6b5a9de46b92fe4edf09a3de6fd65b313b0306f2 but it isn't in the format that @ollieparanoid proposed. I don't really have time this week but next week I should be able to do the things listed (except someone else does them earlier)

**Comment by ollieparanoid** [comment]
Support for this feature has been merged in #356, which also shows how to use it.
I have extended the [deviceinfo reference](https://github.com/postmarketOS/pmbootstrap/wiki/deviceinfo-reference) and added a new sub-article for [deviceinfo_bootimg_qcdt](https://github.com/postmarketOS/pmbootstrap/wiki/deviceinfo_bootimg_qcdt) (which has not much content yet, but it's a start).

The only thing I would like to see improved is saving the `dt.img` file as `dt.img-$FLAVOR`, so we could have multiple kernels installed in parallel. But we can leave that for another PR if this becomes an issue.

Thanks for working on this feature, this is a great enhancement!
